packages/wallet/src/redux/protocols/new-ledger-funding/reducer.ts

```diff
@@ -1,42 +1,385 @@
+  protocolState: states.WaitForPreFundSetup,
+  sharedData: SharedData,
+  action: IDFAction | DirectFundingAction,
+): ReturnVal {
+  if (!isAdvanceChannelAction(action)) {
+    throw new Error(`Incorrect action ${action.type}`);
+  }
+  const preFundResult = advanceChannelReducer(protocolState.preFundSetupState, sharedData, action);
+  sharedData = preFundResult.sharedData;
+  if (!advanceChannelState.isTerminal(preFundResult.protocolState)) {
+    return {
+      protocolState: { ...protocolState, preFundSetupState: preFundResult.protocolState },
+      sharedData,
+    };
+  } else {
+    if (preFundResult.protocolState.type === 'AdvanceChannel.Failure') {
+      return { protocolState: failure({}), sharedData };
+    } else {
+      const ledgerId = preFundResult.protocolState.channelId;
+      const latestCommitment = getLatestCommitment(ledgerId, sharedData);
+
+      const total = latestCommitment.allocation.reduce(addHex);
+      const requiredDeposit = isFirstPlayer(protocolState.channelId, sharedData)
+        ? latestCommitment.allocation[0]
+        : latestCommitment.allocation[1];
+
+      const safeToDepositLevel = isFirstPlayer(protocolState.channelId, sharedData)
+        ? '0x0'
+        : latestCommitment.allocation[1];
+      const ourIndex = isFirstPlayer(protocolState.channelId, sharedData)
+        ? TwoPartyPlayerIndex.A
+        : TwoPartyPlayerIndex.B;
+      // update the state
+      const directFundingAction = directFundingRequested({
+        processId: protocolState.processId,
+        channelId: ledgerId,
+        safeToDepositLevel,
+        totalFundingRequired: total,
+        requiredDeposit,
+        ourIndex,
+      });
+      const directFundingState = initializeDirectFunding(directFundingAction, sharedData);
+      sharedData = directFundingState.sharedData;
+      const advanceChannelResult = initializeAdvanceChannel(
+        protocolState.processId,
+        directFundingState.sharedData,
+        CommitmentType.PostFundSetup,
+        {
+          channelId: ledgerId,
+          ourIndex,
+          processId: protocolState.processId,
+          commitmentType: CommitmentType.PostFundSetup,
+          clearedToSend: false,
+          protocolLocator: NEW_LEDGER_FUNDING_PROTOCOL_LOCATOR,
+        },
+      );
+      sharedData = advanceChannelResult.sharedData;
+      const newProtocolState = states.waitForDirectFunding({
+        ...protocolState,
+        ledgerId,
+        directFundingState: directFundingState.protocolState,
+        postFundSetupState: advanceChannelResult.protocolState,
+      });
+
+      return { protocolState: newProtocolState, sharedData };
+    }
+  }
+}
+
+function handleWaitForDirectFunding(
+  protocolState: states.WaitForDirectFunding,
+  sharedData: SharedData,
+  action: IDFAction | DirectFundingAction | ConsensusUpdateAction,
+): ReturnVal {
+  if (
+    isAdvanceChannelAction(action) &&
+    action.protocolLocator === NEW_LEDGER_FUNDING_PROTOCOL_LOCATOR
```

Comment:
This check seems wrong. The protocolLocator should be used to determine where the action goes, so in this case, the protocolLocator should be `ADVANCE_CHANNEL_PROTOCOL_LOCATOR` if it should be routed to the advanceChannel protocol.

Reply:
Right now we can't just use `ADVANCE_CHANNEL_PROTOCOL_LOCATOR` as the `funding` protocol also uses `advanceChannel` so the `funding` protocol isn't able to determine if an `advanceChannel` action is for itself or `newLedgerFunding`. I think this will be resolved when we get `protocolLocator` working correctly.